devito/ops/transformer.py

```diff
@@ -169,12 +166,18 @@ def create_ops_dat(f, name_to_ops_dat, block):
             ops_dat_array,
             ListInitializer(dat_decls)
         )))
+
+        # Inserting the ops_dat array in case of TimeFunction.
+        name_to_ops_dat[f.name] = ops_dat_array
+
     else:
         ops_dat = OpsDat("%s_dat" % f.name)
         name_to_ops_dat[f.name] = ops_dat
 
-        d_p_val = tuple(sympify([p[0] + h[0] for p, h in zip(f.padding, f.halo)]))
-        d_m_val = tuple(sympify([-(p[1] + h[1]) for p, h in zip(f.padding, f.halo)]))
+        d_p_val = tuple(sympify([p[0] + h[0]
```

Comment:
Thinking about this again... Could you try:

 `d_p_val = tuple(i.left for i in f._size_nodomain)`

or perhaps just:

`d_p_val = f._size_nodoamin.left` 

might work?

about the jargon: nodomain == padding+halo

Reply:
reminder

Reply:
I was trying to use this, but the problem is that the type of the elements inside the tuple of `f._size_nodoamin.left` are `numpy.int64`. Whereas the types in the previous solution are `sympy.core.numbers.Integer`. 
This difference causes error when using the `ListInitializer` (few steps later):
`ValueError: `params` must be an iterable of Expr or str`

And if I remove the `ListInitializer` call I won't have errors but the code generated is wrong:
``` 
u_d_p(d_p) = (0, 2, 2);
# instead of
int u_d_p[2] = {2, 2};
```

So I think I should stick with the current solution.

Reply:
Can u cast them to int(...)

Reply:
I tried casting to `int(...)`, but I get the following error:

```AttributeError: 'int' object has no attribute 'is_Indexed'```

But converting to `sympy.core.numbers.Integer` works!

I just don't know if this is acceptable and if this is better than using `sympfy`, because, in the end, this is what sympy is doing.

Reply:
anyway, clearly this is a deficiency of `ListInitializer`

Turn [this](https://github.com/opesci/devito/blob/master/devito/symbolics/extended_sympy.py#L265) into `is_integer` and we should be good to go leaving everything else as I suggested in the very first place

Reply:
This also is not enough.

So I used the first version, setting `f._size_nodomain.left` as the `d_p_val`.
This property returns a tuple of `numpy.int64`. 

Fixing the `ListInitializer` helped, but the problem is that this is not Sympy object, and later there is pass in

Although now it can identify the `int` inside the `ListInitializer`, it will fail because the type `int` or `np.int64` does not have `is_Indexed` attribute.

The stack trace from the error:
```
The OPS backend is still work-in-progress
Traceback (most recent call last):
  File "04-fun-timefun.py", line 9, in <module>
    op = Operator(Eq(u.forward, u + v.dx))
  File "/home/mickus/Documents/master/projects/devito-vmickus/devito/backends.py", line 73, in __call__
    return t(*args, **kwargs)
  File "/home/mickus/Documents/master/projects/devito-vmickus/devito/ops/operator.py", line 26, in __init__
    super().__init__(*args, **kwargs)
  File "/home/mickus/Documents/master/projects/devito-vmickus/devito/operator.py", line 187, in __init__
    iet = self._specialize_iet(iet, **kwargs)
  File "/home/mickus/Documents/master/projects/devito-vmickus/devito/ops/operator.py", line 70, in _specialize_iet
    pre_time_loop.extend(create_ops_dat(f, name_to_ops_dat, ops_block))
  File "/home/mickus/Documents/master/projects/devito-vmickus/devito/ops/transformer.py", line 196, in create_ops_dat
    res.append(Expression(ClusterizedEq(Eq(d_p, ListInitializer(d_p_val)))))
  File "</home/mickus/miniconda3/envs/devito-vmickus/lib/python3.7/site-packages/decorator.py:decorator-gen-3>", line 2, in __init__
  File "/home/mickus/Documents/master/projects/devito-vmickus/devito/tools/validators.py", line 33, in wrapper
    return f(*args, **kwargs)
  File "/home/mickus/Documents/master/projects/devito-vmickus/devito/ir/iet/nodes.py", line 245, in __init__
    self.__expr_finalize__(expr)
  File "/home/mickus/Documents/master/projects/devito-vmickus/devito/ir/iet/nodes.py", line 250, in __expr_finalize__
    self._reads, _ = detect_io(expr, relax=True)
  File "/home/mickus/Documents/master/projects/devito-vmickus/devito/ir/support/utils.py", line 152, in detect_io
    terminals = flatten(retrieve_terminals(i, deep=True) for i in roots)
  File "/home/mickus/Documents/master/projects/devito-vmickus/devito/tools/utils.py", line 124, in flatten
    for el in l:
  File "/home/mickus/Documents/master/projects/devito-vmickus/devito/ir/support/utils.py", line 152, in <genexpr>
    terminals = flatten(retrieve_terminals(i, deep=True) for i in roots)
  File "/home/mickus/Documents/master/projects/devito-vmickus/devito/symbolics/search.py", line 155, in retrieve_terminals
    return search(expr, q_terminal, mode, 'dfs', deep)
  File "/home/mickus/Documents/master/projects/devito-vmickus/devito/symbolics/search.py", line 114, in search
    return searcher.dfs(expr)
  File "/home/mickus/Documents/master/projects/devito-vmickus/devito/symbolics/search.py", line 67, in dfs
    found.update(self.dfs(a))
  File "/home/mickus/Documents/master/projects/devito-vmickus/devito/symbolics/search.py", line 66, in dfs
    for a in self._next(expr):
  File "/home/mickus/Documents/master/projects/devito-vmickus/devito/symbolics/search.py", line 49, in _next
    if self.deep is True and expr.is_Indexed:
AttributeError: 'numpy.int64' object has no attribute 'is_Indexed'
```

Reply:
can you not convert it to Integer/Number ? In the ListInitializer constructor I mean

so, long story short: there is this `_size_nodomain` method, and we should aim to use it. Accessing `f.padding` or `f.halo` directly doesn't make much sense. In fact, you reminded me that these attributes should be turned into private attributes

Reply:
Sure, can do.